### PR TITLE
[tests] reorganize cache tests and clean memory between tests

### PR DESCRIPTION
**Pull request by gante**
# What does this PR do?

#37394 introduces a cache change, but there are several cache test issues at the moment, preventing a clean merge.

This PR:
1. Removes the `@slow` decorator from a few tests (change pulled from #37394)
2. Deletes `test_static_cache_greedy_decoding_pad_right`, as it doesn't make sense to generate with right padding (change pulled from #37394)
3. Moves `torch.export()` cache tests to its own class
4. Cleans device memory between test runs. On some devices, we were observing CPU weight offloading after a few tests, massively slowing down runtimes. Cleaning the device memory between tests fixes it.

_____________________________________________________
Follow-up PR, before retesting #37394:
1. convert as many `@slow` tests into fast tests as possible
2. fix failing tests

### CODE REVIEW

**Review by zucchini-nlp — APPROVED** [review]
Nice 🧼
